zingo-sync/src/client/fetcher.rs

```diff
@@ -0,0 +1,143 @@
+//! Queue and prioritise fetch requests to fetch data from the server
+
+use std::ops::Range;
+
+use tokio::sync::mpsc::UnboundedReceiver;
+
+use zcash_client_backend::proto::{
+    compact_formats::CompactBlock,
+    service::{
+        compact_tx_streamer_client::CompactTxStreamerClient, BlockId, BlockRange, ChainSpec,
+    },
+};
+use zcash_primitives::consensus::BlockHeight;
+
+use crate::client::FetchRequest;
+
+/// Receives [`self::FetchRequest`]'s via an [`tokio::sync::mpsc::UnboundedReceiver`] for queueing,
+/// prioritisation and fetching from the server.
+/// Returns the data specified in the [`self::FetchRequest`] variant via the provided [`tokio::sync::mpsc::oneshot::Sender`].
+///
+/// Allows all requests to the server to be handled from a single task for efficiency and also enables
+/// request prioritisation for further performance enhancement
+pub async fn fetcher(
+    mut fetch_request_receiver: UnboundedReceiver<FetchRequest>,
+    mut client: CompactTxStreamerClient<zingo_netutils::UnderlyingService>,
+) -> Result<(), ()> {
+    let mut fetch_request_queue: Vec<FetchRequest> = Vec::new();
```

Comment:
Have you thought about using a mcsp::channel directly as the queue? It acts as a fifo list by default so actually works quite well as a queue. Sender threads push directly onto the channel and the fetcher pops them. I think there is some added complexity using a vec as a fifo list (better for filo)..

Reply:
good point, i think we can use a more optimal queue for this reason but it is not designed to be FIFO, we want to prioritise requests. I think your idea about Ordering may be a good solution but as the backpressure should be very low I dont think we need to optimise this yet